File: src/main/java/com/soapdogg/lispInterpreter/printer/ListNotationExpressionNodePrinter.java
package com.soapdogg.lispInterpreter.printer;

import com.soapdogg.lispInterpreter.comparator.NodeValueComparator;
import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants;
import com.soapdogg.lispInterpreter.constants.TokenValueConstants;
import com.soapdogg.lispInterpreter.datamodels.AtomNode;
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.determiner.ExpressionNodeDeterminer;
import lombok.AllArgsConstructor;

@AllArgsConstructor(staticName = "newInstance")
public class ListNotationExpressionNodePrinter {

    private final ExpressionNodeDeterminer expressionNodeDeterminer;
    private final NodeValueComparator nodeValueComparator;
    private final AtomNodePrinter atomNodePrinter;

    String printExpressionNodeInListNotation(
        final ExpressionNode node
    ) {
        Node address = node.getAddress();
        String addressListNotation;
        boolean isAddressList = expressionNodeDeterminer.isExpressionNode(address);
        if (isAddressList) {
            ExpressionNode expressionNodeAddress = (ExpressionNode)address;
            String addressExpressionNodeValue = printExpressionNodeInListNotation(
                expressionNodeAddress
            );
            addressListNotation = TokenValueConstants.OPEN_PARENTHESES + addressExpressionNodeValue;
        } else {
            addressListNotation = atomNodePrinter.printAtomNode((AtomNode)address);
        }

        Node data = node.getData();
        String dataListNotation;
        boolean isDataList = expressionNodeDeterminer.isExpressionNode(data);
        if (isDataList) {
            ExpressionNode expressionNodeData = (ExpressionNode)data;
            String dataExpressionNodeValue = printExpressionNodeInListNotation(
                expressionNodeData
            );
            dataListNotation = ReservedValuesConstants.SPACE + dataExpressionNodeValue;
        } else {
            String dataAtomNodeValue = atomNodePrinter.printAtomNode((AtomNode)data);
            boolean isDataValueNil = nodeValueComparator.equalsNil(dataAtomNodeValue);
            String dataString = isDataValueNil
                ? ReservedValuesConstants.EMPTY
                : ReservedValuesConstants.LIST_DELIMITER + dataAtomNodeValue;

            dataListNotation = dataString + TokenValueConstants.CLOSE_PARENTHESES;
        }
        return addressListNotation + dataListNotation;
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/printer/ListNotationExpressionNodePrinter.kt
package com.soapdogg.lispInterpreter.printer

import com.soapdogg.lispInterpreter.comparator.NodeValueComparator
import com.soapdogg.lispInterpreter.constants.ReservedValuesConstants
import com.soapdogg.lispInterpreter.constants.TokenValueConstants
import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.datamodels.ExpressionNode

class ListNotationExpressionNodePrinter(
    private val nodeValueComparator: NodeValueComparator,
    private val atomNodePrinter: AtomNodePrinter
) {

    fun printExpressionNodeInListNotation(
        node: ExpressionNode
    ): String {
        val addressListNotation = if (node.address is ExpressionNode) {
            val addressExpressionNodeValue = printExpressionNodeInListNotation(
               node.address
            )
            TokenValueConstants.OPEN_PARENTHESES + addressExpressionNodeValue
        } else {
            atomNodePrinter.printAtomNode(node.address as AtomNode)
        }
        val dataListNotation = if (node.data is ExpressionNode) {
            val dataExpressionNodeValue = printExpressionNodeInListNotation(
                node.data
            )
            ReservedValuesConstants.SPACE + dataExpressionNodeValue
        } else {
            val dataAtomNodeValue = atomNodePrinter.printAtomNode(node.data as AtomNode)
            val isDataValueNil = nodeValueComparator.equalsNil(dataAtomNodeValue)
            val dataString = if (isDataValueNil) ReservedValuesConstants.EMPTY else ReservedValuesConstants.LIST_DELIMITER + dataAtomNodeValue
            dataString + TokenValueConstants.CLOSE_PARENTHESES
        }
        return addressListNotation + dataListNotation
    }
}